src/foam/nanos/ruler/RuleEngine.java

```diff
@@ -203,4 +211,61 @@ private void saveHistory(Rule rule, FObject obj) {
     savedRuleHistory_.put(rule.getId(),
       (RuleHistory) ruleHistoryDAO_.put(record).fclone());
   }
+
+  /**
+   * Reloads object when running async action since the original object `obj`
+   * might be stale at the time of execution.
+   *
+   * If reloaded object `nu` hasn't changed (reloaded object equals to old
+   * object) then return the original object as the reloaded object.
+   *
+   * If reloaded object has changed and already saved to DAO (reloaded object
+   * equals to original object) then return the original object.
+   *
+   * Otherwise, return the reloaded object. If greedy flag is set to true then
+   * also copy changes from the original object to the reloaded object.
+   *
+   * @param obj - Original object
+   * @param oldObj - Old object
+   * @param nu - Reloaded object
+   * @param greedy - Flag to set greedy mode
+   * @return Reloaded object
+   */
+  private FObject reloadObject(FObject obj, FObject oldObj, FObject nu, boolean greedy) {
+    FObject old = oldObj;
+    if ( old == null ) {
+      try {
+        old = obj.getClass().newInstance();
+      } catch (Exception e) {
+        // Object instantiation should not fail but if it does fail then return
+        // the original object as the reloaded object.
+        return obj;
+      }
+    }
+
+    old = old.fclone();
```

Comment:
If you put this line in an else clause for the above 'if' you'll save unnecessary cloneing when you create an object on line 238.

Reply:
Oh yeah, you're right.